linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java

```diff
@@ -1297,15 +1297,21 @@ private void closeInner() {
       final Function1<TInner, TKey> innerKeySelector,
       final EqualityComparer<TKey> comparer) {
     return new AbstractEnumerable<TSource>() {
+      private Enumerable<TKey> innerLookup = null;
+      // CALCITE-2909 Delay the computation of the innerLookup until the moment when we are sure
+      // that it will be really needed, i.e. when the first outer enumerator item is processed
+      private Enumerable<TKey> getInnerLookup() {
```

Comment:
I wonder if it makes more sense to create a generic `LazyEnumerable`. So the code becomes :  

```java
return Linq4j.lazy(() -> inner.select(innerKeySelector).distinct());
```
This is similar idea to Guava [Supplier.memoize](https://google.github.io/guava/releases/23.0/api/docs/com/google/common/base/Suppliers.html#memoize-com.google.common.base.Supplier-) or [Observable.defer()](http://reactivex.io/RxJava/2.x/javadoc/io/reactivex/Observable.html#defer-java.util.concurrent.Callable-).

Reply:
I agree, probably it makes sense. I have committed a new version based on a more generic approach.

Reply:
Thank you for `LazyEnumerable` @rubenada. 
Do you mind submitting lazy functionality as separate PR ? I think it makes more sense as stand-alone commit. I will review it and merge as soon as 1.19 release is out.

Reply:
Sure, I have created CALCITE-2937 for that purpose